README.md
# *Hw10Web*

### Console bot with using sqlalchemy models and alembic migrations

[![Language](https://img.shields.io/badge/language-python-blue?&style=plastic)](https://www.python.org)
[![Language version](https://img.shields.io/badge/version-3.10-red?&style=plastic)](https://www.python.org/downloads/)
![GitHub repo size](https://img.shields.io/badge/repo%20size-83%20kB-pink?&style=plastic)

---

### Technology which used:
[![Python](https://img.shields.io/badge/python-3670A0?style=for-the-badge&logo=python&logoColor=ffdd54)](https://www.python.org)
[![Git](https://img.shields.io/badge/git-%23F05033.svg?style=for-the-badge&logo=git&logoColor=white)](https://git-scm.com/)
[![GitHub](https://img.shields.io/badge/github-%23121011.svg?style=for-the-badge&logo=github&logoColor=white)](https://github.com/)
[![MongoDB](https://img.shields.io/badge/MongoDB-%234ea94b.svg?style=for-the-badge&logo=mongodb&logoColor=white)](https://www.mongodb.com/)


---

### Author
[![GitHub Contributors Image](https://contrib.rocks/image?repo=LeadShadow/hw10web)](https://github.com/LeadShadow)

---
### License

[![GitHub](https://img.shields.io/github/license/LeadShadow/hw10web)](https://github.com/LeadShadow/hw10web/blob/main/LICENSE)


history.txt

# 2022-09-14 14:29:29.808130
+quit


src/dml_ab.py
from src.models import Addressbook


def insert_adressbook(name, phone, birthday, emails, address):
    ab = Addressbook(name=name, phone=phone, birthday=birthday, emails=emails, address=address)
    ab.save()
    print(f'Successful add user')


def change_contact(name, old, new):
    r = Addressbook.objects
    for i in r:
        if i.name == name and i.phone == old:
            i.update(phone=new)
    print('Successful change phone number')


def show_phone(name):
    r = Addressbook.objects
    for i in r:
        if i.name == name:
            print(f"{name}'s phone: {i.phone}")


def del_phone(name):
    r = Addressbook.objects
    for i in r:
        if i.name == name:
            i.update(phone='')
    print(f'Successful delete phone')


def show_all():
    r = Addressbook.objects
    if not r:
        print('AddressBook is empty')
    for i in r:
        if i.birthday is not None:
            print(f"User: {i.name}, phone: {i.phone}, birthday: {i.birthday.date()}, email: {i.emails}, address: {i.address}")
        elif i.birthday is None:
            print(f"User: {i.name}, phone: {i.phone}, birthday: {i.birthday}, email: {i.emails}, address: {i.address}")


def add_birthday(name, birthday):
    r = Addressbook.objects
    for i in r:
        if i.name == name and i.birthday is not None:
            print(f'User already has birthday: {i.birthday}')
        if i.name == name and i.birthday is None:
            i.update(birthday=birthday)
    print(f'Successful add birthday to user {name}')


def days_to_user_birthday(name):
    r = Addressbook.objects
    for i in r:
        if i.name == name and i.birthday is not None:
            return i.birthday.date()
        elif i.name == name and i.birthday is None:
            return i.birthday


def find_something(som):
    r = Addressbook.objects
    j_som = ' '.join(som)
    for i in r:
        if j_som in i.name or j_som in i.phone or j_som in i.emails or j_som in i.address:
            print(f"User: {i.name}, phone: {i.phone}, birthday: {i.birthday.date()}, email: {i.emails}, address: {i.address}")


def del_user(name):
    r = Addressbook.objects
    for i in r:
        if i.name == name:
            i.delete()
    print(f'Successful delete user: {name}')


def clear_all():
    r = Addressbook.objects
    for i in r:
        i.delete()
    print('Successful delete all users')


def add_email(name, email):
    r = Addressbook.objects
    for i in r:
        if i.name == name and i.emails != '':
            print('Email already exists')
        if i.name == name and i.emails == '':
            i.update(emails=email)
            print(f'Successful add email to user {name}')


def del_email(name):
    r = Addressbook.objects
    for i in r:
        if i.name == name and i.emails != '':
            i.update(emails='')
            print(f'Successful delete email from user {name}')
        if i.name == name and i.emails == '':
            print('Email does not exist')


def add_address(name, address):
    r = Addressbook.objects
    for i in r:
        if i.name == name and i.address != '':
            print('Email already exists')
        if i.name == name and i.address == '':
            i.update(address=address)
            print(f'Successful add address to user {name}')


if __name__ == "__main__":
    pass


src/models.py
from datetime import datetime
import redis
from redis_lru import RedisLRU
from mongoengine import *

connect(host='mongodb://localhost:27017/test')
client = redis.StrictRedis(host='localhost', port=6379, password=None)
cache = RedisLRU(client)


class Addressbook(Document):
    name = StringField(max_length=100, min_length=1, required=True)
    phone = StringField(max_length=100, default='')
    birthday = DateTimeField(max_length=100, default='', null=True)
    emails = StringField(max_length=100, default='')
    address = StringField(max_length=120, default='')


class Tags(Document):
    id_count = IntField()
    tags = ListField(StringField(max_length=80))


class Note(Document):
    id_count = IntField()
    text = StringField(max_length=1000)
    done = BooleanField(default=False)
    created = DateTimeField(default=datetime.now())
    tags = ReferenceField(Tags, reverse_delete_rule=CASCADE, dbref=True)


class Archived(Document):
    id_count = IntField()
    text = StringField(max_length=1000)
    transferred = DateTimeField(default=datetime.now())
    tags = ReferenceField(Tags, reverse_delete_rule=CASCADE, dbref=True)



src/dml_note.py
from src.models import Note, Archived, Tags, cache


def add_note(text):
    counter = 0
    if not Note.objects and Archived.objects:
        for a in Archived.objects:
            if counter < a.id_count:
                counter = a.id_count
    if not Archived.objects and Note.objects:
        for n in Note.objects:
            if counter < n.id_count:
                counter = n.id_count
    if Archived.objects and Note.objects:
        for a in Archived.objects:
            for n in Note.objects:
                if a.id_count > n.id_count:
                    counter = a.id_count
                else:
                    counter = n.id_count
    if not Note.objects and not Archived.objects:
        counter = 0
    note = Note(id_count=counter + 1,
                text=text)
    note.save()
    print('Successful add note')


def change_note(id_, text):
    notes = Note.objects
    for note in notes:
        if note.id_count == id_:
            note.update(text=text)
            print(f'Successful change note with id: {id_}')


def del_note(id_):
    notes = Note.objects
    for note in notes:
        if note.id_count == id_:
            note.delete()
            print(f'Successful delete note with id: {id_}')


def add_date(id_, date):
    notes = Note.objects
    for note in notes:
        if note.id_count == id_:
            note.update(created=date)
            print(f'Successful add date to note with id: {id_}')


def show_all():
    notes = Note.objects
    for note in notes:
        if note.tags is None:
            print(f'Note id: {note.id_count}, date: {note.created.date()}, done: {note.done}\n'
                  f'Text: {note.text}\n'
                  f'---------------------------------------------------------\n')
        else:
            print(f'Note id: {note.id_count}, date: {note.created.date()}, done: {note.done}\n'
                  f'Text: {note.text}\n'
                  f'Tags: {note.tags.tags}\n'
                  f'---------------------------------------------------------\n')


def done_note(id_):
    notes = Note.objects
    for note in notes:
        if note.id_count == id_:
            arch = Archived(id_count=note.id_count,
                            text=note.text,
                            tags=note.tags)
            arch.save()
            note.delete()
            print(f'Note with id: {id_} added to archive')


def show_archived():
    arch = Archived.objects
    for a in arch:
        if a.tags is None:
            print(f'Note id: {a.id_count}, date: {a.transferred.date()}\n'
                  f'Text: {a.text}\n'
                  f'---------------------------------------------------------\n')
        else:
            print(f'Note id: {a.id_count}, date: {a.transferred.date()}\n'
                  f'Text: {a.text}\n'
                  f'Tags: {a.tags.tags}\n'
                  f'---------------------------------------------------------\n')


def find_note(text):
    notes = Note.objects
    som_st = ' '.join(text)
    for n in notes:
        created = n.created.strftime("%Y-%m-%d %H:%M:%S")
        if som_st in n.text or som_st in created:
            print(f'Note: {n.id_count}, text: {n.text}, created: {n.created}')


def return_note(id_):
    arch = Archived.objects
    for a in arch:
        if a.id_count == id_:
            note = Note(id_count=a.id_count,
                        text=a.text,
                        tags=a.tags)
            note.save()
            a.delete()
            print(f'Successful return note with id: {id_}')


def add_tag(id_, tags):
    list_tags = []
    [list_tags.append(i) for i in tags]
    for n in Note.objects:
        if n.id_count == id_:
            tag = Tags(id_count=id_,
                       tags=list_tags)
            tag.save()
            note = Note(id_count=id_,
                        text=n.text,
                        tags=tag)
            note.save()
            n.delete()
            print(f'Successful add tag/s to note with id: {id_}')


def clear_all():
    notes = Note.objects
    for n in notes:
        n.delete()
        print('Successful delete all notes')


def find_tag(tag):
    notes = Note.objects
    str_tag = ' '.join(tag)
    print(str_tag)
    for n in notes:
        print('1')
        for i in n.tags.tags:
            print(i)
            if str_tag.lower() in i.lower():
                print(f'Note id: {n.id_count}, date: {n.created.date()}, done: {n.done}\n'
                      f'Text: {n.text}\n'
                      f'Tags: {n.tags.tags}\n'
                      f'---------------------------------------------------------\n')


def show_date(date1, date2):
    notes = Note.objects
    for n in notes:
        print(n.created)
        if date1.value <= n.created <= date2.value:
            print(f'Id: {n.id_count}, date: {n.created.date()}')
        else:
            print(f'Not find notes with this date')


if __name__ == "__main__":
    pass


console_bot/notebook.py
"""Модуль для роботи з нотатками"""
from prompt_toolkit import prompt
from prompt_toolkit.history import FileHistory
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.completion import NestedCompleter

from console_bot.command_parser import RainbowLexer
import datetime
import re
import src.dml_note as dml


class DateIsNotValid(Exception):
    """You cannot add an invalid date"""


class InputError:
    """Клас для виклику помилки при введенні невірного даних"""
    def __init__(self, func) -> None:
        self.func = func

    def __call__(self, contacts, *args):
        try:
            return self.func(contacts, *args)
        # except IndexError:
        #     return 'Error! Give me name and phone or birthday please!'
        except KeyError:
            print('Error! Note not found!')
        except ValueError:
            print('Error! Incorrect argument!')
        except DateIsNotValid:
            print('Error! Date is not valid')
        except IndexError:
            print('Error! Incorrect argument!')


class Field:
    def __init__(self, value: str) -> None:
        self.__value = None
        self.value = value

    def __str__(self) -> str:
        return f'{self.value}'

    def __eq__(self, other) -> bool:
        return self.value == other.value

    def __lt__(self, other) -> bool:
        return self.value < other.value

    def __gt__(self, other) -> bool:
        return self.value > other.value

    def __le__(self, other) -> bool:
        return self.value <= other.value

    def __ge__(self, other) -> bool:
        return self.value >= other.value


class ExecDate(Field):
    def __str__(self) -> str:
        if self.value is None:
            return ' - '
        else:
            return f'{self.value:%d %b %Y}'

    @property
    def value(self) -> datetime.date:
        return self.__value

    @value.setter
    def value(self, value: str) -> None:
        if value is None:
            self.__value = None
        else:
            try:
                self.__value = datetime.datetime.strptime(value, '%Y-%m-%d')
            except ValueError:
                raise DateIsNotValid


@InputError
def add_note(*args):
    """Додає нотатку"""
    note_text = ' '.join(args)
    dml.add_note(note_text)


@InputError
def change_note(*args):
    id_note, new_text = int(args[0]), ' '.join(args[1:])
    dml.change_note(id_note, new_text)


@InputError
def del_note(*args):
    id_note = int(args[0])
    yes_no = input(f'Are you sure you want to delete the note ID: {id_note}? (Y/n) ')
    if yes_no == 'Y':
        dml.del_note(id_note)
    else:
        print('Note not deleted')


@InputError
def add_date(*args):
    """Додає дату нотатки"""
    id_note, exec_date = int(args[0]), ExecDate(args[1])
    dml.add_date(id_note, datetime.datetime.strftime(exec_date.value, '%Y-%m-%d'))


def show_all(*args):
    """Повертає всі нотатки"""
    print(dml.show_all())


def show_archive(*args):
    """Повертає нотатки з архіву"""
    dml.show_archived()


def find_note(*args):
    """Повертає нотатки за входженням в текст"""
    dml.find_note(args)


@InputError
def show_date(*args):
    """Повертає нотатки з вказаною датою виконання"""
    date_find = ExecDate(args[0])
    if len(args) > 1:
        days = int(args[1])
    else:
        days = 0
    date1 = (date_find.value - datetime.timedelta(days=days)).strftime('%Y-%m-%d')
    date2 = (date_find.value + datetime.timedelta(days=days)).strftime('%Y-%m-%d')
    dml.show_date(ExecDate(date1), ExecDate(date2))


@InputError
def done_note(*args):
    """Помічає нотатку як виконану"""
    id_note = int(args[0])
    dml.done_note(id_note)


@InputError
def return_note(*args):
    """Помічає нотатку як виконану"""
    id_note = int(args[0])
    dml.return_note(id_note)


@InputError
def add_tag(*args):
    id_note = int(args[0])
    note_tags = re.sub(r'[;,.!?]', ' ', ' '.join(args[1:])).title().split()
    dml.add_tag(id_note, note_tags)


@InputError
def find_tag(*args):
    """Повертає нотатки в яких є тег"""
    dml.find_tag(args)


def goodbye(*args):
    print('You have finished working with notebook0')


def unknown_command(*args):
    print('Unknown command! Enter again!')


def help_me(*args):
    """Повертає допомогу по списку команд"""
    print("""\nCommand format:
    help or ? - this help;
    add note <text> - add note;
    change note <id> <text> - change note;
    delete note <id> - delete note;
    add date <id> <date> - add/change date;
    add tag <id> <tag> - add tag;
    done <id> - mark note as done;
    return <id> - mark note as not done;
    show all - show all notes;
    show archived - show archived notes;
    show date <date> [<days>] - show notes by date +- days;
    find note <text> - find note by text;
    find tag <text> - find note by tag;
    sort by tags - show all notes sorted by tags;
    good bye or close or exit or . - exit the program""")


COMMANDS = {help_me: ['?', 'help'], goodbye: ['good bye', 'close', 'exit', '.'], add_note: ['add note '],
            add_date: ['add date '], show_all: ['show all'], show_archive: ['show archived'],
            change_note: ['change note '], del_note: ['delete note '], find_note: ['find note '],
            show_date: ['show date '], done_note: ['done '], return_note: ['return '], add_tag: ["add tag"],
            find_tag: ["find tag"]}


def command_parser(user_command: str) -> (str, list):
    for key, list_value in COMMANDS.items():
        for value in list_value:
            if user_command.lower().startswith(value):
                args = user_command[len(value):].split()
                return key, args
    else:
        return unknown_command, []


def start_nb():
    print('\n\033[033mWelcome to notebook!\033[0m')
    print(f"\033[032mType command or '?' for help \033[0m\n")
    while True:
        user_command = prompt('Enter command >>> ',
                              history=FileHistory('history.txt'),
                              auto_suggest=AutoSuggestFromHistory(),
                              completer=Completer,
                              lexer=RainbowLexer()
                              )
        command, data = command_parser(user_command)
        command(*data)
        if command is goodbye:
            break


Completer = NestedCompleter.from_nested_dict({'help': None, 'good bye': None, 'exit': None,
                                              'close': None, '?': None, '.': None,
                                              'add': {'note': None, 'date': None, 'tag': None},
                                              'show': {'all': None, 'archived': None, 'date': None},
                                              'change note': None, 'delete note': None,
                                              'find': {'note': None, 'tag': None}, 'done': None,
                                              'return': None, 'sort by tags': None})


if __name__ == '__main__':
    start_nb()


console_bot/file_parser.py
from prompt_toolkit import prompt
from prompt_toolkit.history import FileHistory
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.completion import NestedCompleter
from console_bot.command_parser import RainbowLexer

import shutil
from console_bot.parser import *
from console_bot.normalize import normalize
from console_bot.command_parser import command_parser


def help_me(*args):
    return """\nCommand format:
    help or ? - this help;
    parse folder_name- sorts files in the folder;
    good bye or close or exit or . - exit the program"""


def goodbye(*args):
    return 'You have finished working with file_parser'


def handle_media(filename: Path, target_folder: Path):
    target_folder.mkdir(exist_ok=True, parents=True)
    filename.replace(target_folder / (normalize(filename.name[:-len(filename.suffix)]) + filename.suffix))


def handle_other(filename: Path, target_folder: Path):
    target_folder.mkdir(exist_ok=True, parents=True)
    filename.replace(target_folder / (normalize(filename.name[:-len(filename.suffix)]) + filename.suffix))


def handle_archive(filename: Path, target_folder: Path):
    target_folder.mkdir(exist_ok=True, parents=True)
    folder_for_file = target_folder / normalize(filename.name.replace(filename.suffix, ''))
    folder_for_file.mkdir(exist_ok=True, parents=True)
    try:
        shutil.unpack_archive(str(filename.resolve()), str(folder_for_file.resolve()))
        #     normalize archive files
        for file in folder_for_file.iterdir():
            file.replace(folder_for_file / (normalize(file.name[:-len(file.suffix)]) + file.suffix))

    except shutil.ReadError:
        print(f'Not an archive {filename}!')
        folder_for_file.rmdir()
        return None
    filename.unlink()
    # .resolve gives absolute path


def handle_folder(folder: Path):
    try:
        folder.rmdir()
    except OSError:
        return f"Folder deletion failed {folder}"


def file_parser(*args):
    try:
        folder_for_scan = Path(args[0])
        scan(folder_for_scan.resolve())
    except FileNotFoundError:
        return f"Not able to find '{args[0]}' folder. Please enter a correct folder name."
    except IndexError:
        return "Please enter a folder name."
    for file in JPEG_IMAGES:
        handle_media(file, Path(args[0] + '/' + 'images' + '/' + 'JPEG'))
    for file in JPG_IMAGES:
        handle_media(file, Path(args[0] + '/' + 'images' + '/' + 'JPG'))
    for file in PNG_IMAGES:
        handle_media(file, Path(args[0] + '/' + 'images' + '/' + 'PNG'))
    for file in SVG_IMAGES:
        handle_media(file, Path(args[0] + '/' + 'images' + '/' + 'SVG'))
    for file in MP3_AUDIO:
        handle_media(file, Path(args[0] + '/' + 'audio' + '/' + 'MP3'))
    for file in OGG_AUDIO:
        handle_media(file, Path(args[0] + '/' + 'audio' + '/' + 'OGG'))
    for file in WAV_AUDIO:
        handle_media(file, Path(args[0] + '/' + 'audio' + '/' + 'WAV'))
    for file in AMR_AUDIO:
        handle_media(file, Path(args[0] + '/' + 'audio' + '/' + 'AMR'))
    for file in AVI_VIDEO:
        handle_media(file, Path(args[0] + '/' + 'video' + '/' + 'AVI'))
    for file in MP4_VIDEO:
        handle_media(file, Path(args[0] + '/' + 'video' + '/' + 'MP4'))
    for file in MOV_VIDEO:
        handle_media(file, Path(args[0] + '/' + 'video' + '/' + 'MOV'))
    for file in MKV_VIDEO:
        handle_media(file, Path(args[0] + '/' + 'video' + '/' + 'MKV'))
    for file in DOC_DOCUMENT:
        handle_media(file, Path(args[0] + '/' + 'document' + '/' + 'DOC'))
    for file in DOCX_DOCUMENT:
        handle_media(file, Path(args[0] + '/' + 'document' + '/' + 'DOCX'))
    for file in TXT_DOCUMENT:
        handle_media(file, Path(args[0] + '/' + 'document' + '/' + 'TXT'))
    for file in PDF_DOCUMENT:
        handle_media(file, Path(args[0] + '/' + 'document' + '/' + 'PDF'))
    for file in XLSX_DOCUMENT:
        handle_media(file, Path(args[0] + '/' + 'document' + '/' + 'XLSX'))
    for file in PPTX_DOCUMENT:
        handle_media(file, Path(args[0] + '/' + 'document' + '/' + 'PPTX'))
    for file in OTHER:
        handle_other(file, Path(args[0] + '/' + 'other'))

    for file in ARCHIVES:
        handle_archive(file, Path(args[0] + '/' + 'archives'))

    for folder in FOLDERS[::-1]:
        handle_folder(folder)

    return f"Files in {args[0]} sorted succesffully"


COMMANDS_F = {file_parser: ['parse '], help_me: ['?', 'help'], goodbye: ['good bye', 'close', 'exit', '.']}


def start_fp():
    print('\n\033[033mWelcome to file parser!\033[0m')
    print(f"\033[032mType command or '?' for help \033[0m\n")
    while True:
        with open("history.txt", "wb"):
            pass
        user_command = prompt('Enter command >>> ',
                              history=FileHistory('history.txt'),
                              auto_suggest=AutoSuggestFromHistory(),
                              completer=Completer,
                              lexer=RainbowLexer()
                              )
        command, data = command_parser(user_command, COMMANDS_F)
        print(command(*data), '\n')
        if command is goodbye:
            break


Completer = NestedCompleter.from_nested_dict({'help': None, '?': None, 'parse': None, 'good bye': None,
                                              'close': None, 'exit': None, '.': None})



console_bot/adressbook.py
from prompt_toolkit import prompt
from prompt_toolkit.history import FileHistory
from prompt_toolkit.auto_suggest import AutoSuggestFromHistory
from prompt_toolkit.completion import NestedCompleter
from datetime import datetime
import colorama
import re

from console_bot.command_parser import command_parser, RainbowLexer
import src.dml_ab as dml


class Field:
    def __init__(self, value: str) -> None:
        self.__value = None
        self.value = value

    def __str__(self) -> str:
        return f'{self.value}'


class Name(Field):
    @property
    def value(self):
        return self.__value

    @value.setter
    def value(self, value: str):
        self.__value = value


class Phone(Field):
    @property
    def value(self):
        return self.__value

    @value.setter
    def value(self, value: str):
        def is_code_valid(phone_code: str) -> bool:
            if phone_code[:2] in ('03', '04', '05', '06', '09') and phone_code[2] != '0' and phone_code != '039':
                return True
            return False

        result = None
        phone = value.removeprefix('+').replace('(', '').replace(')', '').replace('-', '')
        if phone.isdigit():
            if phone.startswith('0') and len(phone) == 10 and is_code_valid(phone[:3]):
                result = '+38' + phone
            if phone.startswith('380') and len(phone) == 12 and is_code_valid(phone[2:5]):
                result = '+' + phone
            if 10 <= len(phone) <= 14 and not phone.startswith('0') and not phone.startswith('380'):
                result = '+' + phone
        if result is None:
            raise ValueError(f'Неправильний тип значення {value}')
        self.__value = result


class Birthday(Field):
    def __str__(self):
        if self.value is None:
            return '-'
        else:
            return f'{self.value:%Y-%m-%d}'

    @property
    def value(self):
        return self.__value

    @value.setter
    def value(self, value: str):
        if value is None:
            self.__value = None
        else:
            try:
                self.__value = datetime.strptime(value, '%Y-%m-%d')
            except ValueError:
                raise DateIsNotValid


class Address(Field):
    @property
    def value(self):
        return self.__value

    @value.setter
    def value(self, value: str):
        if value is None:
            self.__value = None
        self.__value = value

    def __str__(self):
        if self.value is None:
            return '-'
        else:
            return f'{self.value}'


class Email(Field):
    @property
    def value(self):
        return self.__value

    @value.setter
    def value(self, value: str):
        if value is None:
            self.__value = None
        else:
            result = None
            get_emails = re.findall(r'\b[a-zA-Z][\w\.]+@[a-zA-Z]+\.[a-zA-Z]{2,}', value)
            if get_emails:
                for i in get_emails:
                    result = i
            if result is None:
                raise AttributeError(f"Неправильний тип значення {value}")
            self.__value = result


def days_to_birthday(birthday: datetime):
    if birthday is None:
        return None
    this_day = datetime.today()
    birthday_day = datetime(this_day.year, birthday.month, birthday.day)
    if birthday_day < this_day:
        birthday_day = datetime(this_day.year + 1, birthday.month, birthday.day)
    return (birthday_day - this_day).days


class PhoneUserAlreadyExists(Exception):
    """You cannot add an existing phone number to a user"""


class EmailUserAlreadyExists(Exception):
    """You cannot add an existing email to a user"""


class DateIsNotValid(Exception):
    """You cannot add an invalid date"""


class EmailIsNotValid(Exception):
    """Email is not valid, try again"""


class FindNotFound(Exception):
    """Find is not valid, try again"""


class InputError:
    def __init__(self, func) -> None:
        self.func = func

    def __call__(self, contacts, *args):
        try:
            return self.func(contacts, *args)
        except IndexError:
            print('Error! Give me name and phone or birthday please!')
        except KeyError:
            print('Error! User not found!')
        except ValueError:
            print('Error! Phone number is incorrect!')
        except PhoneUserAlreadyExists:
            print('Error! You cannot add an existing phone number to a user')
        except EmailUserAlreadyExists:
            print('Error! You cannot add an existing email to a user')
        except DateIsNotValid:
            print('Error! Date is not valid')
        except AttributeError:
            print('Error! Email is not valid')
        except FindNotFound:
            print('Error! Try command find or search "words" that find contact')


def salute(*args):
    print('Hello! How can I help you?')


@InputError
def add_contact(*args):
    name = Name(args[0])
    phone = Phone(args[1])
    birthday = None
    email = ''
    address = ''
    if len(args) > 2:
        birthday = Birthday(args[2])
    if len(args) > 3:
        email = Email(args[3])
    if len(args) > 4:
        address = Address(" ".join(args[4:]))
    if len(args) <= 2:
        birthday = None
    if len(args) <= 3:
        email = ''
    if len(args) <= 4:
        address = ''
    if birthday is None and email == '' and address == '':
        dml.insert_adressbook(name.value, phone.value, birthday, email, address)
    if birthday is not None and email == '' and address == '':
        dml.insert_adressbook(name.value, phone.value, datetime.strftime(birthday.value, '%Y-%m-%d'), email, address)
    if birthday is not None and email != '' and address == '':
        dml.insert_adressbook(name.value, phone.value, datetime.strftime(birthday.value, '%Y-%m-%d'), email.value, address)


@InputError
def change_contact(*args):
    name, old_phone, new_phone = Name(args[0]), Phone(args[1]), Phone(args[2])
    dml.change_contact(name.value, old_phone.value, new_phone.value)


@InputError
def show_phone(*args):
    name = Name(args[0])
    dml.show_phone(name.value)


@InputError
def del_phone(*args):
    name = Name(args[0])
    dml.del_phone(name.value)


def show_all(*args):
    dml.show_all()


@InputError
def add_birthday(*args):
    name, birthday = Name(args[0]), Birthday(args[1])
    dml.add_birthday(name.value, datetime.strftime(birthday.value, '%Y-%m-%d'))


@InputError
def days_to_user_birthday(*args):
    name = Name(args[0])
    birthday = dml.days_to_user_birthday(name.value)
    if birthday is None:
        print('User there is no birthday')
    else:
        print(f'{days_to_birthday(birthday)} days to birthday user {name}')


def goodbye(*args):
    print('You have finished working with addressbook')


@InputError
def search(*args):
    if len(args) >= 1:
        dml.find_something(args)
    else:
        raise FindNotFound


@InputError
def del_user(*args):
    name = Name(args[0])
    yes_no = input(f'Are you sure you want to delete the user {name}? (Y/n) ')
    if yes_no == 'Y':
        dml.del_user(name.value)
    else:
        print('User not deleted')


def clear_all(*args):
    yes_no = input('Are you sure you want to delete all users? (Y/n) ')
    if yes_no == 'Y':
        dml.clear_all()
    else:
        print('Removal canceled')


@InputError
def add_email(*args):
    name, email = Name(args[0]), Email(args[1])
    dml.add_email(name.value, email.value)


@InputError
def del_email(*args):
    name = Name(args[0])
    dml.del_email(name.value)


@InputError
def add_address(*args):
    name, address = Name(args[0]), Address(" ".join(args[1:]))
    dml.add_address(name.value, address.value)


def help_me(*args):
    print("""\nCommand format:
    help or ? - this help;
    hello - greeting;
    add <name> <phone> <birthday> - add user to directory;
    change <name> <old_phone> <new_phone> - change the user's phone number;
    del phone <name> - delete the user's phone number;
    delete <name> - delete the user;
    clear - delete all users;
    birthday <name> <birthday> - add/modify the user's birthday;
    email <name> <email> - add the user's email;
    del email <name> - delete the user's email;
    address <name> <address> - add/modify the user's address;
    show <name> - show the user's data;
    show all - show data of all users;
    find or search <sub> - show data of all users with sub in name, phones or birthday;
    days to birthday <name> - show how many days to the user's birthday;
    good bye or close or exit or . - exit the program""")


COMMANDS_A = {salute: ['hello'], add_contact: ['add '], change_contact: ['change '], help_me: ['?', 'help'],
              show_all: {'show all'}, goodbye: ['good bye', 'close', 'exit', '.'], del_phone: ['del phone '],
              add_birthday: ['birthday'], days_to_user_birthday: ['days to birthday '],
              show_phone: ['show '], search: ['find ', 'search '],
              del_user: ['delete '], clear_all: ['clear'], add_email: ['email '], add_address: ['address'],
              del_email: ['del email']}


def start_ab():
    print('\n\033[033mWelcome to the address book!\033[0m')
    print(f"\033[032mType command or '?' for help \033[0m\n")
    while True:
        user_command = prompt('Enter command >>> ',
                              history=FileHistory('history.txt'),
                              auto_suggest=AutoSuggestFromHistory(),
                              completer=Completer,
                              lexer=RainbowLexer()
                              )
        command, data = command_parser(user_command, COMMANDS_A)
        command(*data), '\n'
        if command is goodbye:
            break


Completer = NestedCompleter.from_nested_dict({'help': None, 'hello': None, 'good bye': None, 'exit': None,
                                              'close': None, '?': None, '.': None, 'birthday': None,
                                              'days to birthday': None, 'add': None,
                                              'show all': None, 'change': None, 'del': {'phone': None, 'email': None}, 'delete': None,
                                              'clear': None, 'email': None, 'find': None, 'search': None,
                                              'address': None})

if __name__ == "__main__":
    start_ab()
